spring-integration-redis/src/main/java/org/springframework/integration/redis/util/RedisLockRegistry.java

```diff
@@ -523,11 +495,108 @@ public boolean equals(Object obj) {
 		private RedisLockRegistry getOuterType() {
 			return RedisLockRegistry.this;
 		}
+	}
+
+	private final class RedisPubSubLock extends RedisLock {
+		private static final String OBTAIN_LOCK_SCRIPT =
+				"local lockClientId = redis.call('GET', KEYS[1])\n" +
+				"if lockClientId == ARGV[1] then\n" +
+				"  redis.call('PEXPIRE', KEYS[1], ARGV[2])\n" +
+				"  return true\n" +
+				"elseif not lockClientId then\n" +
+				"  redis.call('SET', KEYS[1], ARGV[1], 'PX', ARGV[2])\n" +
+				"  return true\n" +
+				"end\n" +
+				"return false";
+
+		private static final String UNLINK_UNLOCK_SCRIPT =
+				"if (redis.call('unlink', KEYS[1]) == 1) then " +
+				"redis.call('publish', ARGV[1], KEYS[1]) " +
+				"return true " +
+				"end " +
+				"return false";
+
+		private static final String DELETE_UNLOCK_SCRIPT =
+				"if (redis.call('del', KEYS[1]) == 1) then " +
+				"redis.call('publish', ARGV[1], KEYS[1]) " +
+				"return true " +
+				"end " +
+				"return false";
+		private final RedisScript<Boolean> obtainLockScript = new DefaultRedisScript<>(OBTAIN_LOCK_SCRIPT, Boolean.class);
+		private final RedisScript<Boolean> unLinkUnLockScript = new DefaultRedisScript<>(UNLINK_UNLOCK_SCRIPT, Boolean.class);
+		private final RedisScript<Boolean> deleteUnLockScript = new DefaultRedisScript<>(DELETE_UNLOCK_SCRIPT, Boolean.class);
+
+		private RedisPubSubLock(String path) {
+			super(path);
+		}
+
+		@Override
+		protected boolean tryRedisLockInner(long time) throws ExecutionException, InterruptedException {
+			return subscribeLock(time);
+		}
+
+		@Override
+		protected void removeLockKeyInnerUnlink() {
+			RedisLockRegistry.this.redisTemplate.execute(
+					this.unLinkUnLockScript, Collections.singletonList(this.lockKey),
+					RedisLockRegistry.this.unLockChannelKey);
+		}
+
+		@Override
+		protected void removeLockKeyInnerDelete() {
+			RedisLockRegistry.this.redisTemplate.execute(
+					this.deleteUnLockScript, Collections.singletonList(this.lockKey),
+					RedisLockRegistry.this.unLockChannelKey);
+
+		}
+
+		private boolean subscribeLock(long time) throws ExecutionException, InterruptedException {
+			final long expiredTime = System.currentTimeMillis() + time;
+			if (obtainLock()) {
+				return true;
+			}
+
+			if (!(RedisLockRegistry.this.isRunningRedisMessageListenerContainer
+				&& RedisLockRegistry.this.redisMessageListenerContainer.isRunning())) {
```

Comment:
I think this `redisMessageListenerContainer` property has to be optional so we don't create extra objects if we don't deal with pub-sub.